internal/controllers/admission_test.go

```diff
@@ -19,159 +19,178 @@ func operator(spec operatorsv1alpha1.OperatorSpec) *operatorsv1alpha1.Operator {
-		}
-	})
-	It("should fail if an invalid channel name is given", func() {
-		invalidChannels := []string{
-			"spaces spaces",
-			"Capitalized",
-			"camelCase",
-			"many/invalid$characters+in_name",
-			"-start-with-hyphen",
-			"end-with-hyphen-",
-			".start-with-period",
-			"end-with-period.",
-		}
-		for _, invalidChannel := range invalidChannels {
-			err := cl.Create(ctx, operator(operatorsv1alpha1.OperatorSpec{
-				PackageName: "package",
-				Channel:     invalidChannel,
-			}))
-			Expect(err).To(HaveOccurred(), "expected error for invalid channel '%q'", invalidChannel)
-			Expect(err.Error()).To(ContainSubstring("spec.channel in body should match '^[a-z0-9]+([\\.-][a-z0-9]+)*$'"))
-		}
-	})
-	It("should pass if a valid channel name is given", func() {
-		validChannels := []string{
-			"hyphenated-name",
-			"dotted.name",
-			"channel-has-version-1.0.1",
-		}
-		for _, validChannel := range validChannels {
-			op := operator(operatorsv1alpha1.OperatorSpec{
-				PackageName: "package",
-				Channel:     validChannel,
-			})
-			err := cl.Create(ctx, op)
-			Expect(err).NotTo(HaveOccurred(), "unexpected error creating valid channel '%q': %w", validChannel, err)
-			err = cl.Delete(ctx, op)
-			Expect(err).NotTo(HaveOccurred(), "unexpected error deleting valid channel '%q': %w", validChannel, err)
-		}
-	})
-	It("should fail if an invalid channel name length", func() {
+		require.Errorf(t, err, "expected error for invalid semver %q", invalidSemver)
+		// Don't need to include the whole regex, this should be enough to match the MasterMinds regex
+		require.ErrorContains(t, err, "spec.version in body should match '^(\\s*(=||!=|>|<|>=|=>|<=|=<|~|~>|\\^)")
+	}
+}
+
+func TestOperatorValidSemver(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	validSemvers := []string{
+		">=1.2.3",
+		"=>1.2.3",
+		">= 1.2.3",
+		">=v1.2.3",
+		">= v1.2.3",
+		"<=1.2.3",
+		"=<1.2.3",
+		"=1.2.3",
+		"!=1.2.3",
+		"<1.2.3",
+		">1.2.3",
+		"~1.2.2",
+		"~>1.2.3",
+		"^1.2.3",
+		"1.2.3",
+		"v1.2.3",
+		"1.x",
+		"1.X",
+		"1.*",
+		"1.2.x",
+		"1.2.X",
+		"1.2.*",
+		">=1.2.3 <2.3.4",
+		">=1.2.3,<2.3.4",
+		">=1.2.3, <2.3.4",
+		"<1.2.3||>2.3.4",
+		"<1.2.3|| >2.3.4",
+		"<1.2.3 ||>2.3.4",
+		"<1.2.3 || >2.3.4",
+		">1.0.0,<1.2.3 || >2.1.0",
+		"<1.2.3-abc >2.3.4-def",
+		"<1.2.3-abc+def >2.3.4-ghi+jkl",
+	}
+	for _, validSemver := range validSemvers {
+		op := operator(operatorsv1alpha1.OperatorSpec{
+			PackageName: "package",
+			Version:     validSemver,
+		})
+		err := cl.Create(ctx, op)
+		require.NoErrorf(t, err, "unexpected error for semver range '%q': %w", validSemver, err)
+		err = cl.Delete(ctx, op)
+		require.NoErrorf(t, err, "unexpected error deleting valid semver '%q': %w", validSemver, err)
+	}
+}
+
+func TestOperatorInvalidChannel(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	invalidChannels := []string{
```

Comment:
General comment for this and the lists of semver ranges above - what are we testing? Is it validating admission? If so - there are straightforward ways to unit-test that logic that will be orders of magnitude faster than this. Lists such as these look like the wrong level of abstraction to test in integration tests.

Reply:
This is a direct translation of the existing tests. I'm not looking to change the nature of the tests here.

Reply:
So, there's a set of validatior tests in a subdirectory of this, to which these can probably be moved.

Reply:
And my recollection is that the complex regex used to validate these semvers may be a bit lenient, so this is a more thorough test.

Reply:
That appears to be the case, not all of these can be easily validated by the admission validation (see #542 https://github.com/operator-framework/operator-controller/pull/542/files#diff-6af2fcdc11cc7d43e85d42a4926bfa2ffd47e2292b09ba20407ba78b0f34c3d1R74), so these need to be left here, as they are testing the parsing code during reconciliation.

Reply:
The regex was sufficiently complex that it was decided to make it a bit more permissive, rather than locking it down and potentially having false negatives. We'd rather have false positives that are caught in the reconcile.

Reply:
Right, so I guess what I am getting at is that this test does not test reconcile ...

```go
			err := cl.Create(ctx, d.spec)
			require.Error(t, err)
			require.ErrorContains(t, err, d.errMsg)
```

Reply:
(and, since it does not, it's just testing admission, which is a test of the regex, and it's therefore superfluous)

Reply:
So, what's the take away from that?

Reply:
Are you saying we should _not_ test a complex regex? Or are you saying there's a better way to test the regex?

Reply:
ping @stevekuznetsov

Reply:
Sorry, missed this - let's keep it for now but open an issue to do it at the unit level. If you follow [this code](https://github.com/operator-framework/operator-lifecycle-manager/blob/7d4f29f923579a3b2363dd031c724139b88b8100/pkg/controller/operators/catalog/operator.go#L2086) you can see how to generate a validator for a structural schema. Then you can  run this a couple orders of magnitude faster and without network hops. I don't think it's good for anyone to run unit-level validations in e2e.

Reply:
A reminder that _this_ is unit tests. The e2e tests are in #545

Reply:
Issue: #549